enclave-tls/samples/enclave-tls-client/client.c

```diff
@@ -49,6 +53,78 @@ static sgx_enclave_id_t load_enclave(bool debug_enclave)
 #endif
 // clang-format on
 
+static uint8_t user_mrsigner[32];
+static uint8_t user_mrenclave[32];
+
+static int parse_hex(const char *hex, void *buffer, size_t buffer_size)
+{
+	if (!hex || !buffer || buffer_size == 0)
+		return -1;
+
+	if (strlen(hex) != buffer_size * 2) {
+		ETLS_ERR("Invalid hex string (%s) length\n", hex);
+		return -1;
+	}
+
+	for (size_t i = 0; i < buffer_size; i++) {
+		if (!isxdigit(hex[i * 2]) || !isxdigit(hex[i * 2 + 1])) {
+			ETLS_ERR("Invalid hex string '%s'\n", hex);
+			return -1;
+		}
+
+		sscanf(hex + i * 2, "%02hhx", &((uint8_t *)buffer)[i]);
+	}
+
+	return 0;
+}
+
+static int parse_config_file(char *path)
+{
+	FILE *fp;
+	char errbuf[200];
+
+	fp = fopen(path, "r");
+	if (!fp) {
+		ETLS_ERR("failed to open /opt/enclave-tls/config.toml\n");
+		goto err;
+	}
+
+	toml_table_t *conf = toml_parse_file(fp, errbuf, sizeof(errbuf));
+	fclose(fp);
+
+	if (!conf) {
+		ETLS_ERR("failed to parse %s\n", errbuf);
+		goto err;
+	}
+	toml_table_t *enclave_info = toml_table_in(conf, "enclave-info");
+	if (!enclave_info) {
+		ETLS_DEBUG("there is no enclave-info data\n");
+		return 0;
+	}
+
+	toml_datum_t mrsigner = toml_string_in(enclave_info, "mrsigner");
+	if (mrsigner.ok) {
+		if (parse_hex(mrsigner.u.s, user_mrsigner, sizeof(user_mrsigner)))
+			return -1;
+	} else
+		ETLS_DEBUG("there is no mrsigner\n");
+
+	toml_datum_t mrenclave = toml_string_in(enclave_info, "mrenclave");
+	if (mrenclave.ok) {
+		if (parse_hex(mrenclave.u.s, user_mrenclave, sizeof(user_mrenclave)))
+			return -1;
+	} else
+		ETLS_DEBUG("there is no mrenclave\n");
```

Comment:
Obviously what we really want to implement is https://github.com/alibaba/inclavare-containers/issues/967 addresses. Because etls is a general cross-platform implementation, it is impossible to implement the parser of  hw-tee specific parameters in core layer. Instead, the verifier instance should register a config parser method to core layer and the global config parser in core layer will call back the verifier instance to parse the configs specific to it. This design will decouple core layer and verifier instance.

Reply:
In addition, please split this pr into 3 commits:

1. Implement https://github.com/alibaba/inclavare-containers/issues/657 (global config file)
2. Implement #967 (allow to call user_callback and hw-tee specific config parser)
3. Implement the general user_callback mechanism
4. Implement a simple sgx evidence check for mrenclave and mrsigner and the reference values retrieved from global config file.

In future, we only need to re-use the above infrastructure and implement OPA check.